ambry-frontend/src/main/java/com/github/ambry/frontend/DeleteDatasetHandler.java

```diff
@@ -136,7 +143,20 @@ private void deleteDataset() throws RestServiceException {
         accountName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_ACCOUNT_NAME, true);
         containerName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_CONTAINER_NAME, true);
         datasetName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_DATASET_NAME, true);
-        //TODO: delete all the dataset version under the dataset.
+        List<DatasetVersionRecord>
+            datasetVersionRecordList = accountService.getAllValidVersion(accountName, containerName, datasetName);
+        for (DatasetVersionRecord record : datasetVersionRecordList) {
+          String version = record.getVersion();
+          RequestPath requestPath = getRequestPath(restRequest);
+          RequestPath newRequestPath =
+              new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                  NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
+                      + version, requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+          // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+          restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+          restRequest.setArg(Headers.DATASET_VERSION_QUERY_ENABLED, "true");
+          deleteBlobHandler.handle(restRequest, restResponseChannel, null);
+        }
         accountService.deleteDataset(accountName, containerName, datasetName);
```

Comment:
how many dataset versions are we talking about here? wil this take a long time to delete?

Reply:
Usually it won't be a lot, because we will support auto retention logic as well. But worst case if might have 100 blobs. This deleteDataset call is not critical, so if it failed, customer would take best effort to retry.